### ISSUE TITLE
Kirbytext: add basic srcset support to (image:)

### ISSUE THREAD

**Issue by fvsch** [getkirby-v2/kirby#73]
This is a suggestion for an enhancement. I’m not sure it would be worthwhile or what the scope of that feature should be, so I’m opening this issue to discuss it.

The idea is to add basic and “automatic” `srcset` support to the `<img>` element generated by kirbytext with the `(image: …)` syntax.

By automatic I mean based on the available files, using naming conventions. That seems to be the Kirby Way. :)

Spec reference for image candidate descriptors:
https://html.spec.whatwg.org/multipage/embedded-content.html#image-candidate-string
## Browser support

I suggest only outputting a correct `srcset`, and living it up to themes to include a polyfill for srcset if they want to.
## Pixel density

```
(image: awesome-cat.jpg alt: On this picture you can see a cat)
```

with those files:

```
./awesome-cat.jpg
./awesome-cat@1.5x.jpg
./awesome-cat@2x.jpg
```

should output:

``` html
<img src="awesome-cat.jpg" alt="On this picture you can see a cat"
    srcset="awesome-cat@1.5x.jpg 1.5x, awesome-cat@2x.jpg 2x">
```

(I skipped the full paths to the content folder, for readability’s sake.)
## Width

```
./awesome-cat.jpg
./awesome-cat@500w.jpg
./awesome-cat@1280w.jpg
```

should output:

``` html
<img src="awesome-cat.jpg" alt="On this picture you can see a cat"
    srcset="awesome-cat@500w.jpg 500w, awesome-cat@1280w.jpg 1280w">
```

Warning: I’m not sure the width descriptors are any use without a `sizes` attribute on the `<img>`.
## With both density and width descriptors

Maybe that’s too complex to be useful, plus the way I read the spec it’s supposed to be illegal though as a use case it makes sense to me. -__-
I need to sort this out.

Anyway:

```
./awesome-cat.jpg
./awesome-cat@1x500w.jpg
./awesome-cat@2x500w.jpg   (1000px wide)
./awesome-cat@1x1280w.jpg
./awesome-cat@2x1280w.jpg  (2560px wide)
```

should output:

``` html
<img src="awesome-cat.jpg" alt="On this picture you can see a cat"
  srcset="awesome-cat@1x500w.jpg 1x 500w,
          awesome-cat@2x500w.jpg 2x 500w,
          awesome-cat@1x1280w.jpg 1x 1280w,
          awesome-cat@2x1280w.jpg 2x 1280w">
```
## Specified source has descriptors

```
(image: awesome-cat@1x500w.jpg)
```

we can still output:

``` html
<img src="awesome-cat@1x500w.jpg" alt="On this picture you can see a cat"
  srcset="awesome-cat@1x500w.jpg 1x 500w,
          <!-- other candidates if any -->">
```

Just mentioning this because it means the work to be done when checking for alternative versions of the image should include stripping the descriptor from the image’s filename (if any), _then_ matching against other filenames.
## Limit to pixel density?

Maybe the pixel density use case is the only one that makes sense without specifying a `sizes` attribute with complex values (to be determined, sorry for the lack of research). So perhaps this enhancement should be limited to the pixel density descriptors. But it doesn’t hurt — and shouldn’t take much more work — having support for detecting the width descriptors and outputting them.

**Comment by erikcollinder** [comment]
+1

**Comment by Designscheu** [comment]
+1

**Comment by philgruneich** [comment]
I was thinking about this lately, considered the following:
````
(image: awesome-cat.jpg alt: On this picture you can see a cat srcset=awesome-cat@1.5x.jpg 1.5x, awesome-cat@2x.jpg 2x)
````
I prefer to insert the files on the kirbytag instead of forcing users into a filename format. I believe this requires only 1 foreach loop (or a really smart regex) over the srcset tag to grab the absolute url for the image. As usual, I could be wrong.

**Comment by jancbeck** [comment]
@philgruneich kirbytag syntax would 
(image: awesome-cat.jpg alt: On this picture you can see a cat srcset:awesome-cat@1.5x.jpg 1.5x, awesome-cat@2x.jpg 2x)

**Comment by philgruneich** [comment]
@jancbeck Oops, altered my example (:

**Comment by jancbeck** [comment]
To get this thing going, I submitted a pull request. 

In my solution you can basically put whatever you want into the srcset attribute as long as the the filename is referenced followed by a space. Just follow the spec, make sure the referenced file actually exists in the active page directory and you are good to go. It supports as many descriptors as you like.

**Comment by philgruneich** [comment]
That's terrific Jan. Thank you.

I think that's the most flexible approach to this issue.

**Comment by fvsch** [comment]
`<opinion>`

Meh. I stick by the "detect multiple sources from existing files following a naming convention", which is the Kirby Way imo. People who want total freedom can use the `<img>` tag.

`</opinion>`

**Comment by jancbeck** [comment]
@fvsch I'll try to create an implementation that follows your approach. Ultimately I don't see why we shouldn't support both.

**Comment by jancbeck** [comment]
@bastianallgeier any thoughts on this?

**Comment by philgruneich** [comment]
@fvsch I think the biggest benefit of the image kirbytag is grabbing the absolute urls for you, so it is not a case of *use `<img>` instead*. What other naming convention Kirby has that affects Kirbytext, so we can put this in perspective?

**Comment by bastianallgeier** [comment]
I wasn't really able so far to follow the picture vs. srcset war closely enough, but I'm still a bit hesitant to add this to the core. I wonder if it shouldn't be a plugin. You can easily overwrite the existing image tag from within a plugin. So all you'd need to do is to install such a plugin and get started with the modified version. I'd feel better with that until it's finally clear which way to go. What do you think?

**Comment by malvese** [comment]
There are many different ways to use srcset (and picture). Is there a most common case for when an (image: …) is used?
My opinion is not every image needs to be "retina". On the other hand smaller sources are a must for devices on cellular networks and for responsive layouts.

I also think most editors would prefer to upload a single image and let the CMS do its magic, without having to specify sources and sizes for every image.
We could create different sources on the fly with thumb(), then use whatever we want in the templates or controllers.

Perhaps we can go crazy with plugins for now, and if a particular implementation is gaining traction it can be included in core.

**Comment by fvsch** [comment]
@bastianallgeier I agree that it’s better to see one or a few implementations with plugins first, and maybe pick an approach for Kirby later (if there’s value in that), rather than risking adding bloat and/or an untested approach now.

For the record the picture/srcset thing is settled (it’s srcset mostly, and picture+srcset for a limited set of advanced use cases). What’s pretty much in flux is figuring out what makes sense for users, both site managers (esp. CMS users) and site visitors.

**Comment by philgruneich** [comment]
I agree with the "let's build a plugin first" approach.

**Comment by jancbeck** [comment]
I agree too.

**Comment by diondiondion** [comment]
@bastianallgeier The war between different RespImg implementations has been over for quite some time now; the specification has been agreed upon, some browsers have implemented it and there's a working polyfill.

That aside, I'd take Brad Frost's advice on the matter to heart: http://bradfrost.com/blog/post/responsive-images/

> So here’s how I’m going to write 95% of my responsive images:
> ´< img src="small.jpg" srcset="small.jpg [smallwidth]w, large.jpg [largewidth]w" alt="Alt Text" />´
> That’s it.
> My biggest priority is avoiding sending gigantic images to small screens, and this does the trick.

..I feel like this should be relatively easy to implement in a Kirby-esque way.

**Comment by philgruneich** [comment]
@diondiondion I don't think that's the point. Besides, only Chrome currently supports the sizes parameter for srcset without any flags on. Also, I have a scenario where I actually need both resolution and sizes, so I'd like that syntax, so despite agreeing with Brad's advice, there are specific cases.

**Comment by malvese** [comment]
@philgruneich Well, support will improve with time. I don't think we need a plugin that covers every possible case, it would be very complex. And I don't see the point, it's easy enough with Kirby to treat specific cases in a template or any other way.

**Comment by jancbeck** [comment]
I like the idea to let kirby handle all the things in the background so the user has to do nothing expect linking an image. A few thoughts about where this is problematic though and needs a clever solution:

### Pixel density
I often take screenshots on my Retina MacBook that are always at @2x. For people without retina devices it doesn't make sense to show them a 2x version and if I don't limit the size manually it will be shown at twice the normal size. So here the background script should jump in and generate a 1x version. Here it would make sense to enforce a naming scheme like `$filename@2x.$extension`. I see no way around this since there is no way of telling the pixel density programatically. 
In another use case I share pictures taken with my camera. I always upload images in their highest resolution and want people to see a 2x version if their device supports it. Here the background script would generate a 2x version if it finds no @2x marker in the filename. 
However then again consider someone who takes screenshots on their non-retina device. The screenshot would show only half the size on 2x devices. The only way I see out of it is enforcing a @x naming scheme for all images. Said user would have to name their file screenshot@1x.png and I would have to name mine photo@2x.jpg. 
I think the **user** should decide on a **file-by-file** basis how pixel density is handled. Perhaps there could be an option how a situation without a filename-marker would be handled and a default.

### Breakpoints
A common use case is where you have photo or screenshot that is 2000px wide, a few hundred kilobytes heavy and you don't want a feature phone on a EDGE connection squeeze that to its 320px wide display.
In that case breakpoints would make sense like small:480px, medium:768px, large:1024px. Who should decide about these breakpoints? The plugin developer/kirby core, the template developer or the user? I think the user should have a say in this via an option but there should be a default value based on common use cases.

### Actual display size

For mobile devices with small screens my template renders everything in one column at 100vw. For higher resolutions I show a fixed-size sidebar and some padding around the main content column. For cases like this, there is the sizes attribute that supports the calc() function. Something like `sizes="(min-width: 30em) 100vw - 300px - (2 * 10vw ), 100vw"` would describe the actual width of the image to the browser in any situation. 
I think the user shouldn't have to mess with any of this. Instead the template provider should figure this out and set an option in a format that we agree upon and the responsive image plugin should use that or use smart default. 
Compare this to how WordPress allows theme developers to set a global [$content_width](https://codex.wordpress.org/Content_Width) variable.

**Comment by diondiondion** [comment]
> Here it would make sense to enforce a naming scheme like $filename@2x.$extension. I see no way around this since there is no way of telling the pixel density programatically.

But of course there is - photos normally have their dpi embedded as meta information. Even MacBook screenshots are correctly tagged with their dpi value that can be used to make a decision.

Breakpoints and the end-user display size of images should be down to the creator of the templates, I think; the kirby config file seems like a good place to store this information.

It would be useful to be able to define different sets of sizes (for example "sidebar-image", "article-image", etc) since an image in a column would have to be sized differently than one that spans the whole screen. I'm not sure how these "sizing" classes should then be assigned to images, though. Maybe the most straightforward way would be to bind them to html classes, so that when an image is added to a page with a kirbytag and a class, like so: (image: file.jpg caption: text class: full-width), the corresponding sizes attribute values will be added to the generated img tag automatically.

**Comment by jancbeck** [comment]
> But of course there is - photos normally have their dpi embedded as meta information. 
Any examples of how to get this information in PHP? Parsing PNG files is way above my pay grade.

> Breakpoints and the end-user display size of images should be down to the creator of the templates, I think; the kirby config file seems like a good place to store this information.

Agreed. Something along the lines of 

c::set('responsiveimages.breakpoints', array( 480, 768, 1024 ));
c::set('responsiveimages.sizes', array( '(min-width: 30em) 100vw - 300px - (2 * 10vw )', '100vw' ));
c::set('responsiveimages.quality', 50;

> It would be useful to be able to define different sets of sizes (for example "sidebar-image", "article-image", etc) since an image in a column would have to be sized differently than one that spans the whole screen.

That can be done by any template developer already. No need for a plugin here. 

I believe this plugin should have its focus on providing an easy solution for content creators. That means the main article space, not sidebars and secondary columns that totally depend on a template. Perhaps in the beging there will be a custom kirbytext tag like `(responsive-image: filename.jpg)` that will be a fork of the native `(image)` tag that could be merged back into core at one point. Or it will just act as a standalone. Using a pre-filter, all instances of the native image tag could be dynamically replaced by the responsive one. That way users wouldn't have to change their content. I'm a bit hesitant to the idea of simply overwriting kirbytext::$tags['image'].

**Comment by malvese** [comment]
> But of course there is - photos normally have their dpi embedded as meta information. Even MacBook screenshots are correctly tagged with their dpi value that can be used to make a decision.

@diondiondion I believe DPIs don't matter on a computer display, they are only useful when you print a picture. Displays' DPIs vary widely. Besides, DPIS don't work in every browser (Chrome, Safari…) and use device-pixel-ratio instead.

> I often take screenshots on my Retina MacBook that are always at @2x.

@jancbeck I think it is outside of the plugin's scope, it's an edge case better handled through the template's code or a custom field. There are tools like http://www.retinacapture.com/ to take screenshots at both "Retina" and "normal" resolution.
I don't think it is possible for a plugin to guess when an image should be halved.

**Comment by JimmyRittenborg** [comment]
Please be aware that `sizes` isn't optional anymore, when using `srcset`

>Why?

>Too many people were relying on the invisible default for images whose display size wasn’t even close to 100vw, resulting in inappropriate srcset selections and wasted bytes. Developers learning the new markup were also getting tripped up on how sizes affects intrinsic sizing.

>_Source: http://alistapart.com/blog/post/article-update-dont-rely-on-default-sizes_

**Comment by malvese** [comment]
TL;DR
The way I see it:
- Keep the usual Kibrytext image tag, *don't* add new attributes for responsive purposes ;
- if image derivatives are available, the plugin uses them ;
- if not, it creates them when possible ;
- it uses custom breakpoints for ```sizes``` ;
- default breakpoints if none is configured ;
- for complex and edge cases, use the ```<img>``` element.

---

In his opening post, @fvsch wrote "The idea is to add basic and “automatic” srcset support to the <img> element generated by kirbytext with the (image: …) syntax."
By that, I believe he didn't mean we should add optional attributes to (image: …), he meant that people should keep using the Kirby tag as usual, and the plugin would use the appropriate images derivatives.

We should always try to make editors' work as easy as possible, and fiddling with filenames and Kirby tag attributes is not a good path towards that goal. Typing an ```srcset``` for every single image in your content doesn't help at all, it's more work.
I should add than in my opinion the primary goal of responsive images is serving the right images for the display size, to speed page display. Serving hi-res, higher quality ones is a worthy but secondary goal.

I also agree with @jancbeck that this is aimed primarily at images in a main content field. Images with their own field or in sidebars should be dealt with in templates.

**Re-reading the whole discussion**, I see two possible scenarios:
- The editor uploads a single image. Preferably the biggest he has. In this case the plugin should generate derivative images in multiple sizes. My opinion is that this should be done under the hood ; the derivatives should not appear in the page's sidebar, otherwise if the image must be replaced the editor would have to delete/replace all of them.  
The thumb() function seems to be a good candidate for this task (or any contributed replacement like [Image converter](https://github.com/madergrafisch/kirby-imageconverter)) and would follow a naming scheme. For now let's pretend we create 3 derivatives and append '_small', '_medium', or '_large' to the file name.
- The editor uploads several images, following the exact naming convention described above. The plugin would then use these user-created images instead of creating them.  
Not sure what should be done if the editor only uploads 2 images ('_small' and '_large') instead of 3, should the plugin try to create the '_medium' one?  
To be honest I see very few reasons why people would do this manually instead of letting the plugin do its magic: carefully optimized images maybe? A different cropping while keeping the same h/w ratio (otherwise the <picture> element would be needed)?  
The whole thing seems error prone and puts the burden on the editor.

Obviously my preference is the first scenario, a single image and automatic derivatives, and if both are possible I expect the automatic one to be the most widely used, by a mile.

**Now, how do we approach it?**

The whole point of picture/srcset is that browsers will choose the most appropriate image, among the ones we allow them to use. We don't hand-pick the best one, it is chosen according to a bunch of rules we know nothing about. Well, we know a bit, but ultimately the browser has the final word.
So a 1920px image could be used as @x2 in a 768px or less viewport, as a @x3 in a 640px one. That is, unless the browser decides that since we're on a phone in an area with spotty coverage, a fast download of the standard @x1 image is better than a slow high-res one.
That's the beauty of it, in most cases we don't need to deal with pixel density, no need to give every image a specific one.

Another reason not to worry about pixel densities is that specifying them like this:
```html
<img src="standard.jpg" srcset="retina.jpg 2x, fovea.jpg 3x" />
```
only works with fixed width images, when their display width doesn't change. Yet, in responsive sites the content area/column is almost always fluid, with images sized accordingly (img {max-width: 100%}). This would fetch large hi-res images in narrow contexts and waste bandwidth.

**So we're left with ```srcset``` and ```sizes```.** Every site is different, but very often the content area/column and the viewport have a different width. Even on phones there is margin/padding around the content so the text doesn't extend flush to the sides, so the content width is almost never 100vw. For better results (no bandwidth waste) we obviously need to explicitly tell the plugin how it should best use our images. 
@jancbeck already wrote about how we could add this configuration to the Kirby config so I won't dive into the details for now, but it seems to be a good way to achieve this.
We could even go further and specify breakpoints for specific blueprints, as they don't necessarily all share the same layout.

What if we don't configure our responsive images at all? Then we can apply sensible defaults. I'm not sure what sizes would work best, perhaps we can take hints from the most used breakpoints (around 480, 640, 768, 990, 1200, according to a study I can't find again right now). We probably don't need many derivatives, 2 or 3 sizes would be enough to speed up the pages.
The point here is that a set of images is always better than a single big fixed-width one. Default ```sizes```are better than nothin'!

**Comment by jancbeck** [comment]
> Re-reading the whole discussion, I see two possible scenarios:
> [...]
> Obviously my preference is the first scenario, a single image and automatic derivatives, and if both are implemented I expect the automatic one to be the most widely used, by a mile.

They don't need to be *separate* scenarios though. A plugin could work perfectly, *without any configuration* whatsoever, leaving anything to sensible defaults and generating everything automatically. *If*, however, the user *would* provide options, attributes or extra files, the plugin would happily pick them up, catering for a better over all experience. At one point these decisions - breakpoints, sizes, pixel density, file selection - have to be made, and why not let the user have a say into it? Heck, the image tag already supports the `rel` attribute. Who ever needs that, right?

> - for complex and edge cases, use the `<img>` element.

I assumed we are using the `<img>` element for all of this. Am I missing something?

> So a 1920px image could be used as @x2 in a 768px or less viewport, as a @x3 in a 640px one. That is, unless the browser decides that since we're on a phone in an area with spotty coverage, a fast download of the standard @x1 image is better than a slow high-res one.
> That's the beauty of it, in most cases we don't need to deal with pixel density, no need to give every image a specific one.

I could live with that!

> we could add this configuration to the Kirby config so I won't dive into the details for now, but it seems to be a good way to achieve this.
> We could even go further and specify breakpoints for specific blueprints, as they don't necessarily all share the same layout.
I haven't tested this, but the config values are pretty much changeable from everywhere, right? So I could set the breakpoints, say in /site/templates/grid.php to a different value than in /site/templates/article.php. As long as it happens before the kirbytext has been parsed, that would give me different results, depending on the used template.

*** 

Lot's of good ideas here! I think this is gonna be great.

**Comment by malvese** [comment]
> They don't need to be mutually exclusive scenarios though.
Yeah, that's what I meant, sorry my sentence wasn't clear: if both are implemented *in the plugin* I expect to use the fully automatic mode most of the time.

> I assumed we are using the <img> element for all of this. Am I missing something?
Here again I wasn't clear. Options are fine, I just don't see the point in replicating every ```<picture>``` attribute in a Kirby tag, that's why I believe we'd rather type a whole html image element –in html, in the textarea– instead of a Kirbytext tag with a dozen attributes.

I'll rephrase in my original post, so future reader will hopefully not being confused :)

> I haven't tested this, but the config values are pretty much changeable from everywhere, right?
Ooh yes. Then no need to fill the config file with settings for a single blueprint. We can put it in the template, where we use it. Nice.

**Comment by jancbeck** [comment]
Thanks for clearing that up @malvese. Let's see who comes up with something and we can discuss some actual code.

**Comment by diondiondion** [comment]
> @diondiondion I believe DPIs don't matter on a computer display, they are only useful when you print a picture.

That's their original purpose, but it doesn't change the fact that they hold the information we care about in the scenario brought up by jancbeck, which is the ratio of pixels to physical size. It doesn't matter whether the physical size is a screen or a printed document, neither does browser support (since we'd only use it server-side). But anyhow - I agree that we probably don't need to deal with pixel density here at all.

> That can be done by any template developer already. No need for a plugin here.
> I believe this plugin should have its focus on providing an easy solution for content creators. That means the main article space, not sidebars and secondary columns that totally depend on a template.

> I also agree with @jancbeck that this is aimed primarily at images in a main content field. Images with their own field or in sidebars should be dealt with in templates.

In every kirby project I've built so far I've allowed the content creators to add classes to images in their main content area to show images either full-width, or floated left or right at a smaller size ([example](http://www.possumslog.com/travel/london/alternative-london-tour)), so I don't really consider this an edge case... It would be nice to be able to assign classes to different sets of config if it isn't too much of a hassle to implement.

**Comment by malvese** [comment]
@diondiondion Yes, you're right. We can read the dpi info, and it still can be used by @jancbeck as his need is very specific. But I've found it to be not reliable at all. Cameras in particular seem to use a wide range of dpis, and why not? A photo has no dpi, it's an arbitrary value slapped there for printing reasons. Just last night I was sent 9 jpegs for a project, and found 4 different values among them: 180, 314, 150 and 300 dpi.
The gist of it is that for general photo handling with the plugin, we can't rely on dpis.

Re/ floated images in the main content area, yes ideally there should be a way to apply specific image settings to them. It would save a lot of bandwidth.
Since we can already add classes in Kirbytext image tags, we can use a similar 'class' attribute. If the goal it to create a plugin that could replace the core one someday, I think it's safe to match all the existing attributes.
We could then override the general settings with others for images with specific classes:

```js
c::set('responsiveimages.sizes',
    array( '(min-width: 30em) 100vw - (2 * 10vw )', '100vw' )
);

overridden by:

c::set('responsiveimages.sizes.iDontKnowHowThisWorks',
    array( '(min-width: 30em) 100vw - (2 * 10vw )', '25vw' ),
    array('left', 'right', 'someOtherClass') // applies to these classes
);
```

**Comment by jancbeck** [comment]
Just make it associative and test for a non-numeric key in the array:

``` php
// applies always
c::set('responsiveimages.sizes', array( 
    '(min-width: 30em) 100vw - (2 * 10vw )', '100vw' 
));

// applies only when image class matches keyname
c::set('responsiveimages.sizes', array( 
    'left'       => array( '(min-width: 30em) 100vw - (2 * 10vw )', '25vw' ),
    'full-width' => array( '(min-width: 30em) 100vw - (2 * 10vw )', '100vw' )
));
```

**Comment by diondiondion** [comment]
That would be a beautiful solution!

**Comment by malvese** [comment]
I wrote the classes in an array, because I think quite often several classes can share the same settings (like similar images, the only difference being a float to the left or right). But we can just repeat the sizes array when that happens.
I think it's easier for editors to use a single class ("left") than two or more ("left" + "half" or whatever if the width is half the content width).

**Comment by jancbeck** [comment]
Hello everyone,

## What we can learn from WordPress

although it has been a bit quiet here, the issue has been brought back into my focus, because of my other ego's involvement with WordPress development, where [responsive images are scheduled to be supported by core soon](https://make.wordpress.org/core/2015/08/25/responsive-image-support-update/). Their goals are different from Kirbys philosophy, but we can learn from it nevertheless. 
Here is the gist of it:
- The Responsive Issues Community Group (RICG) has been working on a feature plugin to test the viability of adding responsive image support natively into WordPress. That plugin is on [Github](https://github.com/ResponsiveImagesCG/wp-tevko-responsive-images) and in the [WordPress repository](https://wordpress.org/plugins/ricg-responsive-images/) for everyone to play with.
- They focused on a "invisible" introduction of responsive image support for users. That means no options and no change in workflows - it just works.
- For developers, there are some options available and also some handy functions that help in theme and plugin development.
- Optional advanced image compression (via ImageMagick)
- Optional picturefill.js polyfill

## Kirby Responsive Images Plugin

I then created a first draft for a responsive image plugin for Kirby and I'd be happy to get some feedback from testers. 
Github: https://github.com/jancbeck/kirby-responsive-images
Demo (sort of): http://jancbeck.com/articles/huffduff-safari

Simply download it and move it to the plugins directory and all your images created via [kirbytext](http://getkirby.com/docs/cheatsheet/kirbytags/image) should have srcset and sizes rules attached to them.

Use the config to overwrite sizes:

```php
c::set('responsiveimages.sizes', array( 
    array(
		'size_value' => '10em',
		'mq_value'   => '60em',
		'mq_name'    => 'min-width'
    ),
    array(
		'size_value' => '$img_width',
		'mq_value'   => '30em',
		'mq_name'    => 'min-width'
    ),
    'alignleft' => array(
		'size_value' => 'calc(50vw - 30px)'
    ),
));
```
You guessed correctly: $img_width is a variable that will be replaced by the images native size. The item with the 'alignleft' key is only rendered if the same class is also used in the imgclass attribute. 
Note that the sizes attribute only affects the display of the image object. Think of it as a `width` attribute on steroids. You can use media queries, calc and viewport-percentage lengths to cover any use case. If you want to force the native size of the image, you can use the width attribute like you would do normally with kirbytext:

    (image:workflow@3x.jpg link:workflow@3x.jpg width:1244)

The image has a native width of 3730 pixels (taken on an iPhone 6+) and is shrunk to 1244 pixels (~3x) so it won't blow up on non-retina devices. Other than that there is no support for retina, since browsers are clever enough to figure out what image sources they need load for optimal display. Chrome on a Retina MacBook will load the bigger files than it would on a non-Retina MacBook. The latter would not load the 3730 pixel version of the file but one of the smaller versions generated by Kirby.

How do I change what sizes that are *generated* by Kirby? Use a config like this:

```php
c::set('responsiveimages.sources', array( 
    'small'  => array('width' => 444),
    'medium' => array('width' => 666),
    'large'  => array('width' => 999, 'grayscale' => true),
));
```
The item keys (small, medium...) are meaningless at the moment and not used anywhere. Each array item however is directly passed to the [thumb](http://getkirby.com/docs/cheatsheet/helpers/thumb) function. That means you can use all the parameters that thumb supports too, such as `quality` or `crop`.

## Template functions

I have borrowed a little bit of code from WP's implementation. The sizes syntax is from them and concept of separating the sizes and srcset attributes too. You could use the `kirby_get_srcset` / `kirby_get_sizes` and `kirby_get_srcset_array` / `kirby_get_sizes_array` functions to get srcset attribute values in your template or snippets. However, I don't really want to encourage anyone to do this, because this API is probably going to change in the future.

Ideally, we could have a syntax that hooks into Kirbys chain, like this:

```php
<?php foreach($page->images()->sortBy('sort', 'asc') as $image): ?>
      <figure>
        <img src="<?php echo $image->url() ?>" srcset="<?php $echo $image->srcset() ?>" sizes="<?php $echo $image->sizes() ?>" alt="<?php echo $page->title()->html() ?>">
      </figure>
<?php endforeach ?>
```
But I don't think we can have that without official support from Kirby core :cry: 

## Todo
I'm pretty happy so far and there isn't much I'd add:
- [Advanced Image Compression](http://www.smashingmagazine.com/2015/06/efficient-image-resizing-with-imagemagick/) are custom settings for Image Magick that produce 82% smaller files compared to the default settings without noticeable loss of quality. The instructions are a bit above my head though, and could be material for a plugin on its own that registers a custom image driver.
- Perhaps we could consider image height too and not just width, if there is demand for it.
- The aforementioned chaining syntax a la `$image->srcset()` if core supports it

## What next
I hope as many people as possible give this plugin a try and I'm very happy about suggestions or even pull requests.

**Comment by h0r0man** [comment]
It’s easy to create a custom tag, just copy and modify [kirby image tag](https://github.com/getkirby/kirby/blob/411a62308329b3f4ac68789cbab80767ad184da6/extensions/tags.php#L60) to your needs.

**Comment by h0r0man** [comment]
For more like: image optimization and cache, use instead [PageSpeed Module](https://developers.google.com/speed/pagespeed/module/).

**Comment by dionysiusmarquis** [comment]
As a hybrid solution its also possible to add the requirements as comments. I now browserfy Is already in many eco systems, but i just wanted to leave this here as a fast to implement variant :) because you wouldn't have to change anything. You just need to add a comment on top of every file. 

https://www.npmjs.com/package/gulp-resolve-dependencies

**Comment by maxfenton** [comment]
@jancbeck THANK YOU!

**Comment by catlan** [comment]
Here a simple image2x tag impl. It use _ instead of the  @ in the filename.

### site/tags/image2x.php:

```
<?php

kirbytext::$tags['image2x'] = array(
  'attr' => array(
    'width',
    'height',
    'alt',
    'text',
    'title',
    'class',
    'imgclass',
    'linkclass',
    'caption',
    'link',
    'target',
    'popup',
    'rel',
    'figure'
  ),
  'html' => function($tag) {

    $url     = $tag->attr('image2x');
    $alt     = $tag->attr('alt');
    $title   = $tag->attr('title');
    $link    = $tag->attr('link');
    $caption = $tag->attr('caption');
    $file    = $tag->file($url);
    $srcset  = $tag->attr('srcset');

    $useFigure = (kirby()->option('kirbytext.image.figure') or !empty($caption)) && empty($figure);
    if ($tag->attr('figure') == 'false')
      $useFigure = false;

    // use the file url if available and otherwise the given url
    $url = $file ? $file->url() : url($url);

    // alt is just an alternative for text
    if($text = $tag->attr('text')) $alt = $text;

    // try to get the title from the image object and use it as alt text
    if($file) {

      if(empty($alt) and $file->alt() != '') {
        $alt = $file->alt();
      }

      if(empty($title) and $file->title() != '') {
        $title = $file->title();
      }
      
      $retinaFilename = $file->name() . '_2x.' . $file->extension();
      $retinaFile = $file->page()->file($retinaFilename);
      if (empty($srcset) and $retinaFile and $retinaFile->exists()) {
        $srcset = $retinaFile->url() . ' 2x';
      }

    }

    // at least some accessibility for the image
    if(empty($alt)) $alt = ' ';

    // link builder
    $_link = function($image) use($tag, $url, $link, $file) {

      if(empty($link)) return $image;

      // build the href for the link
      if($link == 'self') {
        $href = $url;
      } else if($file and $link == $file->filename()) {
        $href = $file->url();
      } else if($tag->file($link)) {
        $href = $tag->file($link)->url();
      } else {
        $href = $link;
      }

      return html::a(url($href), $image, array(
        'rel'    => $tag->attr('rel'),
        'class'  => $tag->attr('linkclass'),
        'title'  => $tag->attr('title'),
        'target' => $tag->target()
      ));

    };

    // image builder
    $_image = function($class) use($tag, $url, $alt, $title, $srcset) {
      return html::img($url, array(
        'srcset' => $srcset,
        'width'  => $tag->attr('width'),
        'height' => $tag->attr('height'),
        'class'  => $class,
        'title'  => $title,
        'alt'    => $alt
      ));
    };

    if($useFigure) {
      $image  = $_link($_image($tag->attr('imgclass')));
      $figure = new Brick('figure');
      $figure->addClass($tag->attr('class'));
      $figure->append($image);
      if(!empty($caption)) {
        $figure->append('<figcaption>' . html($caption) . '</figcaption>');
      }
      return $figure;
    } else {
      $class = trim($tag->attr('class') . ' ' . $tag->attr('imgclass'));
      return $_link($_image($class));
    }

  }
);
```

### site/plug-ins/site-methods.php:

```
Site::$methods['image2x'] = function(Site $site, $path='', $style='', $alt='') {
   if (!is_string($path) or $path == '') {
      return '';
   }

   $path_parts = pathinfo($path);

   // Retrieve timestamp, convert to base36 and only keep the last 5 chars,
   // which cover the last 700 days (more than enough for HTTP caches).
   $file = kirby()->roots()->assets() . DS . $path;
   $time = file_exists($file) ? (string) filemtime($file) : '';
   if ($time) $time = substr(base_convert((string) $time, 10, 36), -5);

   $file2x = kirby()->roots()->assets() . DS . $path_parts['dirname'] . DS . $path_parts['filename'] . '_2x.' . $path_parts['extension'];
   $time2x = file_exists($file2x) ? (string) filemtime($file2x) : '';
   if ($time2x) $time2x = substr(base_convert((string) $time2x, 10, 36), -5);

   // Another, and theoretically better option is to put the timestamp
   // in the filename, e.g. myfile-6rwtq.css, but that would require changing
   // the .htaccess to rewrite those requests (stripping the timestamp).
   $base = kirby()->urls()->assets() . '/' . $path;
   $base2x = str_replace($path_parts['basename'], $path_parts['filename'] . '_2x.' . $path_parts['extension'], $base);

   $src = $time ? "$base?v=$time" : $base;
   $srcset = file_exists($file2x) ? "$base2x?v=$time 2x" : "";

   return '<img src="'. $src .'" srcset="'. $srcset .'" style="'. $style .'" alt="'.$file2x.'" />';
};
```

**Comment by bastianallgeier** [comment]
This is now a v3 feature.
